{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import sys\n",
    "import re\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from mapper import expand, parse_mapping_table, apply_mappers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Section 1  (Refenece data setup)\n",
    "## Load Onco Gene Data (Reference Information)\n",
    "### The data below is a curated set of genes known to transform a cell into a tumor cell, or a gene known to suppress"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### OncoKB (Very popular source for cancer gene annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Isoform</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>Entrez Gene ID</th>\n",
       "      <th>Hugo Symbol</th>\n",
       "      <th>Is Oncogene</th>\n",
       "      <th>Is Tumor Suppressor Gene</th>\n",
       "      <th>Highest Level of Evidence(sensitivity)</th>\n",
       "      <th>Highest Level of Evidence(resistance)</th>\n",
       "      <th>Summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000318560</td>\n",
       "      <td>NM_005157.4</td>\n",
       "      <td>25</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "      <td>R1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000502732</td>\n",
       "      <td>NM_007314.3</td>\n",
       "      <td>27</td>\n",
       "      <td>ABL2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000331925</td>\n",
       "      <td>NM_001199954.1</td>\n",
       "      <td>71</td>\n",
       "      <td>ACTG1</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000263640</td>\n",
       "      <td>NM_001111067.2</td>\n",
       "      <td>90</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ENST00000220592</td>\n",
       "      <td>NM_012154.3</td>\n",
       "      <td>27161</td>\n",
       "      <td>AGO2</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Isoform          RefSeq  Entrez Gene ID Hugo Symbol Is Oncogene  \\\n",
       "0  ENST00000318560     NM_005157.4              25        ABL1         Yes   \n",
       "1  ENST00000502732     NM_007314.3              27        ABL2         Yes   \n",
       "2  ENST00000331925  NM_001199954.1              71       ACTG1          No   \n",
       "3  ENST00000263640  NM_001111067.2              90       ACVR1         Yes   \n",
       "4  ENST00000220592     NM_012154.3           27161        AGO2          No   \n",
       "\n",
       "  Is Tumor Suppressor Gene Highest Level of Evidence(sensitivity)  \\\n",
       "0                       No                                      1   \n",
       "1                       No                                    NaN   \n",
       "2                      Yes                                    NaN   \n",
       "3                       No                                    NaN   \n",
       "4                       No                                    NaN   \n",
       "\n",
       "  Highest Level of Evidence(resistance)  Summary  \n",
       "0                                    R1      NaN  \n",
       "1                                   NaN      NaN  \n",
       "2                                   NaN      NaN  \n",
       "3                                   NaN      NaN  \n",
       "4                                   NaN      NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# from http://oncokb.org/#/cancerGenes\n",
    "OncoKB_mut = pd.read_csv(\"../data/allAnnotatedVariants.txt\",sep =\"\\t\", encoding='latin-1')\n",
    "OncoKB_gtype = pd.read_csv(\"../data/allCuratedGenes.txt\",sep =\"\\t\")\n",
    "\n",
    "OncoKB_gtype.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Split out genes into category buckets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OncoKB OG: 276 TSG: 285 OG_TSG: 33\n"
     ]
    }
   ],
   "source": [
    "\n",
    "og_genes = []\n",
    "tsg_genes = []\n",
    "og_tsg_genes = [] \n",
    "\n",
    "for index, row in OncoKB_gtype.iterrows():\n",
    "    if row['Is Oncogene'] == 'Yes':\n",
    "        og_genes.append(row['Hugo Symbol'])\n",
    "    \n",
    "    if row['Is Tumor Suppressor Gene'] == 'Yes':\n",
    "        tsg_genes.append(row['Hugo Symbol'])\n",
    "        \n",
    "    if row['Is Oncogene'] =='Yes' and row['Is Tumor Suppressor Gene'] == 'Yes':\n",
    "        og_tsg_genes.append(row['Hugo Symbol'])\n",
    "\n",
    "OncoKB = {\"OG\":set(og_genes),\"TSG\": set(tsg_genes),\"OG_TSG\": set(og_tsg_genes)}\n",
    "\n",
    "\n",
    "print(\"OncoKB OG:\",len(OncoKB[\"OG\"]),\"TSG:\",len(OncoKB[\"TSG\"]),\"OG_TSG:\",len(OncoKB[\"OG_TSG\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bailye (Another source of cancer genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bailey_2018 OG: 115 TSG: 162 both: 18\n"
     ]
    }
   ],
   "source": [
    "bailye = {\"OG\":[],\"TSG\":[],\"OG_TSG\":[]}\n",
    "bailye_drivers = pd.read_excel(\"../data/mmc1.xlsx\", sheet_name=\"Table S1\", \n",
    "                               skiprows = 3)\n",
    "bailye_drivers = bailye_drivers[[\"Gene\",\"Tumor suppressor or oncogene prediction (by 20/20+)\"]]\n",
    "#print(bailye_drivers.shape[0])\n",
    "bailye_drivers.drop_duplicates(inplace=True)\n",
    "bailye_drivers.columns = [\"gene\", \"role\"]\n",
    "#print(bailye_drivers.shape[0])\n",
    "bailye_drivers.fillna(\"NA\", inplace = True)\n",
    "bailye_drivers[\"role\"] = bailye_drivers[\"role\"].apply(lambda x: x.replace(\"possible \",\"\"))\n",
    "bailye_drivers.drop_duplicates(inplace=True)\n",
    "#print(bailye_drivers.shape[0])\n",
    "bailye_drivers.loc[bailye_drivers.duplicated(\"gene\", keep=False),:]\n",
    "#print(len(set(bailye_drivers[\"gene\"].values)))\n",
    "bailye[\"OG\"] = bailye_drivers.loc[bailye_drivers[\"role\"] == \"oncogene\",:]\n",
    "bailye[\"OG\"] = set(bailye[\"OG\"][\"gene\"].values)\n",
    "bailye[\"TSG\"] = bailye_drivers.loc[bailye_drivers[\"role\"] == \"tsg\",:]\n",
    "bailye[\"TSG\"] = set(bailye[\"TSG\"][\"gene\"].values)\n",
    "bailye[\"OG_TSG\"] = bailye[\"OG\"].intersection(bailye[\"TSG\"])\n",
    "print(\"Bailey_2018 OG:\",len(bailye[\"OG\"]),\"TSG:\",len(bailye[\"TSG\"]),\"both:\", len(bailye[\"OG_TSG\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## NCBI\n",
    "Below loads another set of standard reference genes for the human genome. These are all the genes, independent on if they are releated to cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ok: no empty rows detected\n",
      "Ok: no duplicated pairs detected\n",
      "Ok: All Symbol rows are not empty.\n",
      "Ok: All Symbol are mapped to GeneID\n",
      "13 Symbol mapped to multiple GeneID\n",
      "Ok: All GeneID are unique\n",
      "59665 Symbol can be mapped directly to GeneID\n"
     ]
    }
   ],
   "source": [
    "NCBI = pd.read_csv(\"../data/Homo_sapiens.gene_info\",sep = \"\\t\")\n",
    "NCBI = NCBI[[\"#tax_id\",\"GeneID\",\"Symbol\",\"Synonyms\",\"type_of_gene\"]]\n",
    "NCBI = NCBI.loc[NCBI[\"#tax_id\"] == 9606]\n",
    "NCBI = NCBI.loc[NCBI[\"type_of_gene\"] != \"unknown\"]\n",
    "ncbi_symbols = parse_mapping_table(NCBI, \"Symbol\",\"GeneID\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Helper Methods\n",
    "\n",
    "Methods below are used to score the impact that the mutations / variants from the WES file have on an individual gene. This allows us to represent all the mutations into a value that can be compared across cell lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene_id - Entrez Gene ID \n",
    "# df - all mutations in sample\n",
    "#    columns: \"aa_pos\",\"function\", \"Entrez_gene_ID\"\n",
    "# hotspots - table of known hotsport  \"Entrez_gene_ID\" and \"aa_pos\"\n",
    "\n",
    "def distance_to_hotspot(gene_id,df_gene, hotspots):\n",
    "    '''Returns minimal distance between point mutation in gene and known hotspot'''\n",
    "    # consider only GoF because LoF affect the whole protein and not a certian point\n",
    "    df_gene = df_gene.loc[df_gene[\"function\"] == \"point_mutation\"]\n",
    "    df_hotspots = hotspots.loc[hotspots[\"Entrez_gene_ID\"] == gene_id, :]\n",
    "    dist = 100000\n",
    "    for s,e in set(df_gene[\"aa_pos\"].values):\n",
    "        for hs, he in set(df_hotspots[\"aa_pos\"].values):\n",
    "            dist = min(dist, abs(s-hs),abs(s-he), abs(e-hs), abs(e-he))\n",
    "    return dist\n",
    "\n",
    "def mutations2score(df,TSG,OG,hotspots, w=0.05,verbose = False):\n",
    "    # define gene type\n",
    "    gene_scores = {}\n",
    "    for gene_id in list(set(df[\"Entrez_gene_ID\"].values)):\n",
    "        df_gene = df.loc[df[\"Entrez_gene_ID\"] == gene_id, :]\n",
    "        # functions of mutations\n",
    "        mut_functions = set(df_gene[\"function\"].values)\n",
    "        if gene_id in TSG:\n",
    "            # check for truncating mutations:\n",
    "            # score is -1 if any\n",
    "            if \"truncating_mutation\" in mut_functions:\n",
    "                gene_scores[gene_id] =-1\n",
    "            else:\n",
    "                # in mutation is not truncating, \n",
    "                # get the distance to closets hotspot if any \n",
    "                dist = distance_to_hotspot(gene_id, df_gene, hotspots)\n",
    "                # set abs. maximal of 1/(dist+1) and 2w which corresponds LoF in unannotatedd gene\n",
    "                gene_scores[gene_id] = -max(1.0/(dist+1),2*w)\n",
    "        elif gene_id in OG:\n",
    "            if \"point_mutation\" in mut_functions:\n",
    "                dist = distance_to_hotspot(gene_id, df_gene, hotspots)\n",
    "                if \"truncating_mutation\" in mut_functions:\n",
    "                    if verbose:\n",
    "                        print(\"GoF at dist\",dist,\"from hotspot in Oncogene with GoF\",file = sys.stderr)\n",
    "                        print(df_gene,file = sys.stderr)\n",
    "                    if 1.0/(dist+1) > 2*w:\n",
    "                        # if there is a mutation close to hotspot\n",
    "                        # despite LoF it is likely a driver and the second copy of this gene is broken\n",
    "                        # this happens in sampl 905968  in PIK3CA gene (5290)\n",
    "                        gene_scores[gene_id] = 1.0/(dist+1)\n",
    "                    else:\n",
    "                        # otherwise we set -1 to differentiate it from neutral mutations\n",
    "                        # because oncogene might turn on after domain truncation\n",
    "                        gene_scores[gene_id] = -1\n",
    "                else:\n",
    "                    # non-hotspot mutation in non-LoF oncogene - small positive value\n",
    "                    gene_scores[gene_id] = max(1.0/(dist+1),w)\n",
    "            else:\n",
    "                # only LoF mutations, oncogene is likely not functional or\n",
    "                # it is turned-on after truncting mutations\n",
    "                # therefore we set -1 to differentiate it from neutral mutations\n",
    "                gene_scores[gene_id] = -1\n",
    "            \n",
    "            \n",
    "        else: # if unknown_significance\n",
    "            if \"truncating_mutation\" in mut_functions:\n",
    "                gene_scores[gene_id] = -2.0*w\n",
    "            else:\n",
    "                gene_scores[gene_id] = -w\n",
    "    \n",
    "    return pd.Series(gene_scores)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutation / Hotspot Reference Sources\n",
    "\n",
    "Here we are trying to understand the volatility of the gene by the mutations that are within them. These are outlined in the data files from other data sources similar to OncoKB "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1D mutations 3004 495\n",
      "1D hotspots 1024 55\n",
      "3D mutations 9512\n",
      "3D hotspots  3395\n",
      "genes with hotspot 688\n",
      "unique hotspots 4258\n"
     ]
    }
   ],
   "source": [
    "hotspots2D_subst = pd.read_excel(\"../data/hotspots_v2.xls\",\"SNV-hotspots\")\n",
    "hotspots2D_indels = pd.read_excel(\"../data/hotspots_v2.xls\",\"INDEL-hotspots\")\n",
    "print(\"1D mutations\",hotspots2D_subst.shape[0],hotspots2D_indels.shape[0] )\n",
    "hotspots2D_subst = hotspots2D_subst.loc[~hotspots2D_subst[\"Amino_Acid_Position\"].str.contains(\"splice\")]\n",
    "hotspots2D_subst = hotspots2D_subst[[\"Hugo_Symbol\",\"Amino_Acid_Position\"]]\n",
    "hotspots2D_subst.drop_duplicates(inplace=True)\n",
    "hotspots2D_indels = hotspots2D_indels[[\"Hugo_Symbol\",\"Amino_Acid_Position\"]]\n",
    "hotspots2D_indels.drop_duplicates(inplace=True)\n",
    "print(\"1D hotspots\",hotspots2D_subst.shape[0],hotspots2D_indels.shape[0] )\n",
    "cBioPortal_muts = pd.concat([hotspots2D_subst,hotspots2D_indels],axis =0)\n",
    "cBioPortal_muts_ = cBioPortal_muts.loc[cBioPortal_muts[\"Amino_Acid_Position\"].str.contains(\"-\"),:]\n",
    "cBioPortal_muts = cBioPortal_muts.loc[~cBioPortal_muts[\"Amino_Acid_Position\"].str.contains(\"-\"),:]\n",
    "cBioPortal_muts[\"Amino_Acid_Position\"] = cBioPortal_muts[\"Amino_Acid_Position\"].apply(int)\n",
    "#cBioPortal_muts[\"start\"] = cBioPortal_muts[\"Amino_Acid_Position\"]\n",
    "#cBioPortal_muts[\"end\"] = cBioPortal_muts[\"Amino_Acid_Position\"]\n",
    "cBioPortal_muts[\"aa_pos\"] = cBioPortal_muts[\"Amino_Acid_Position\"].apply(lambda x : (x,x))\n",
    "\n",
    "#cBioPortal_muts_ [\"start\"] = cBioPortal_muts_[\"Amino_Acid_Position\"].apply(lambda x: int(x.split(\"-\")[0]))\n",
    "#cBioPortal_muts_[\"end\"] = cBioPortal_muts_[\"Amino_Acid_Position\"].apply(lambda x: int(x.split(\"-\")[1]))\n",
    "cBioPortal_muts_[\"aa_pos\"] = cBioPortal_muts_[\"Amino_Acid_Position\"].apply(lambda x : tuple(map(int,x.split(\"-\"))))\n",
    "\n",
    "cBioPortal_muts = pd.concat([cBioPortal_muts_,cBioPortal_muts],axis = 0)\n",
    "cBioPortal_muts[\"source\"] = \"1D\"\n",
    "\n",
    "\n",
    "### 3D\n",
    "\n",
    "cBioPortal_3D = pd.read_excel(\"../data/3d_hotspots.xls\",\"Table S5\")\n",
    "cBioPortal_3D = cBioPortal_3D[[\"Gene\",\"Amino_Acid_Position\"]]\n",
    "print(\"3D mutations\",cBioPortal_3D.shape[0])\n",
    "cBioPortal_3D.drop_duplicates(inplace=True)\n",
    "print(\"3D hotspots \",cBioPortal_3D.shape[0])\n",
    "cBioPortal_3D[\"Amino_Acid_Position\"] = cBioPortal_3D[\"Amino_Acid_Position\"].apply(int)\n",
    "#cBioPortal_3D[\"start\"] = cBioPortal_3D[\"Amino_Acid_Position\"]\n",
    "#cBioPortal_3D[\"end\"] = cBioPortal_3D[\"Amino_Acid_Position\"]\n",
    "cBioPortal_3D[\"aa_pos\"] = cBioPortal_3D[\"Amino_Acid_Position\"].apply(lambda x : (x,x))\n",
    "cBioPortal_3D.rename({\"Gene\":'Hugo_Symbol'}, axis=\"columns\",inplace=True)\n",
    "cBioPortal_3D[\"source\"] = \"3D\"\n",
    "cBioPortal_muts = cBioPortal_muts[[\"Hugo_Symbol\",\"aa_pos\"]]\n",
    "cBioPortal_3D = cBioPortal_3D[[\"Hugo_Symbol\",\"aa_pos\"]]\n",
    "cBioPortal_muts = pd.concat([cBioPortal_3D, cBioPortal_muts],axis=0)\n",
    "cBioPortal_muts.sort_values([\"Hugo_Symbol\"],ascending = True,inplace=True)\n",
    "\n",
    "cBioPortal_genes = set(cBioPortal_muts[\"Hugo_Symbol\"].values)\n",
    "print(\"genes with hotspot\",len(cBioPortal_genes))\n",
    "cBioPortal_muts = cBioPortal_muts[[\"Hugo_Symbol\",\"aa_pos\"]]\n",
    "cBioPortal_muts.drop_duplicates(inplace=True)\n",
    "print(\"unique hotspots\",cBioPortal_muts.shape[0])\n",
    "#cBioPortal_muts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compare mutaitons with onco genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes with and without mutations\n",
      "OG oncokb 117 / 276 Bailye_2018 81 / 115\n",
      "TSG oncokb 130 / 285 Bailye_2018 81 / 162\n",
      "OG_TSG oncokb 16 / 33 Bailye_2018 14 / 18\n"
     ]
    }
   ],
   "source": [
    "print(\"genes with and without mutations\")\n",
    "for kind in [\"OG\",\"TSG\",\"OG_TSG\"]:\n",
    "    oncokb = OncoKB[kind].intersection(cBioPortal_genes) \n",
    "    b = bailye[kind].intersection(cBioPortal_genes)\n",
    "    print(kind,\"oncokb\", len(oncokb),\"/\",len(OncoKB[kind]),\n",
    "         \"Bailye_2018\", len(b),\"/\",len(bailye[kind]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OG oncokb 276 Bailye_2018 115 union 335 mutated: 148\n",
      "TSG oncokb 285 Bailye_2018 162 union 351 mutated: 142\n",
      "OG_TSG oncokb 33 Bailye_2018 18 union 49 mutated: 29\n"
     ]
    }
   ],
   "source": [
    "for kind in [\"OG\",\"TSG\",\"OG_TSG\"]:\n",
    "    oncokb = OncoKB[kind]#.intersection(cBioPortal_genes) \n",
    "    b = bailye[kind]#.intersection(cBioPortal_genes)\n",
    "    print(kind,\"oncokb\", len(oncokb),\n",
    "         \"Bailye_2018\", len(b),\"union\",len(b.union(oncokb)),\n",
    "         \"mutated:\",\n",
    "         len(b.union(oncokb).intersection(cBioPortal_genes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TSG: 351 with hotspots: 142 \n",
      "OG with hotspots: 148\n",
      "Hotspots (all from cBioPortal): 4258 \n",
      "\ton OG: 1128 \n",
      "\ton TSG: 1056\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>aa_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1456</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>(206, 206)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1307</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>(258, 258)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1128</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>(328, 328)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2542</td>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>(398, 398)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2056</td>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>(401, 401)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Hugo_Symbol      aa_pos\n",
       "1456       ACVR1  (206, 206)\n",
       "1307       ACVR1  (258, 258)\n",
       "1128       ACVR1  (328, 328)\n",
       "2542      ACVR1B  (398, 398)\n",
       "2056      ACVR1B  (401, 401)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TSG - not necessarity with a hotspot, take union of OncoKB and  Bailye_2018\n",
    "TSG = OncoKB[\"TSG\"].union(bailye[\"TSG\"])\n",
    "# OG - must have a hotspot \n",
    "OG = OncoKB[\"OG\"].union(bailye[\"OG\"]).intersection(cBioPortal_genes)\n",
    "print(\"TSG:\",len(TSG),\"with hotspots:\",len(TSG.intersection(cBioPortal_genes)),\n",
    "      \"\\nOG with hotspots:\",len(OG))\n",
    "# list of hotspots: \n",
    "hotspots = cBioPortal_muts.loc[cBioPortal_muts[\"Hugo_Symbol\"].isin(OG|TSG),:]\n",
    "print(\"Hotspots (all from cBioPortal):\", cBioPortal_muts.shape[0],\n",
    "      \"\\n\\ton OG:\", hotspots.loc[hotspots[\"Hugo_Symbol\"].isin(OG)].shape[0],\n",
    "      \"\\n\\ton TSG:\", hotspots.loc[hotspots[\"Hugo_Symbol\"].isin(TSG)].shape[0])\n",
    "hotspots.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Often gene come with multiple names - This attempts to clean this up\n",
    "\n",
    "At the end we will have all genes translated to have\n",
    "\n",
    "Entrez Gene Id\n",
    "\n",
    "```Entrez Gene maintains records from genomes which have been completely sequenced, which have an active research community to submit gene-specific information, or which are scheduled for intense sequence analysis```\n",
    "\n",
    "Hugo Symbol\n",
    "\n",
    "``` The HUGO Gene Nomenclature Committee is a committee of the Human Genome Organisation that sets the standards for human gene nomenclature. ```\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ok: no empty rows detected\n",
      "Ok: no duplicated pairs detected\n",
      "Ok: All Synonyms rows are not empty.\n",
      "Ok: All Synonyms are mapped to GeneID\n",
      "3198 Synonyms mapped to multiple GeneID\n",
      "50201 different Synonyms mapped to the same GeneID\n",
      "10893 Synonyms can be mapped directly to GeneID\n"
     ]
    }
   ],
   "source": [
    "ncbi_synonyms = expand(NCBI[[\"Synonyms\",\"GeneID\"]],column=\"Synonyms\",sep=\"|\") \n",
    "ncbi_synonyms = parse_mapping_table(ncbi_synonyms, \"Synonyms\",\"GeneID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped: 458 \n",
      "\tdirectly via main_mapper 446 \n",
      "\tvia alternative mapper 2 \n",
      "\tvia one of multiple synonyms in alternative mapper 10 \n",
      "\tLOC 0 \n",
      "Unmapped: 0 \n",
      "\trecognized symbols without Entrez ID 0 \n",
      "\tmultiple query_ids map to the same target_id 0 \n",
      "\tquery_ids map to multiple target_ids in the main mapper 0 \n",
      "\tquery_ids map to multiple target_ids in the alternative mapper 0 \n",
      "\tLOC not found in Entrez 0 \n",
      "\tNot found at all: 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IDs mapped to multiple target IDs are kept:\n",
      " []\n",
      "/Users/kbloom/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \"\"\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>aa_pos</th>\n",
       "      <th>Entrez_gene_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1456</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>(206, 206)</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1307</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>(258, 258)</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1128</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>(328, 328)</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2542</td>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>(398, 398)</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2056</td>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>(401, 401)</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Hugo_Symbol      aa_pos  Entrez_gene_ID\n",
       "1456       ACVR1  (206, 206)              90\n",
       "1307       ACVR1  (258, 258)              90\n",
       "1128       ACVR1  (328, 328)              90\n",
       "2542      ACVR1B  (398, 398)              91\n",
       "2056      ACVR1B  (401, 401)              91"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tsg_og_gene_names = pd.DataFrame.from_dict({\"genes\":dict(zip(list(OG | TSG),list(OG | TSG)))})\n",
    "tsg_og_gene_names, query2target, not_mapped = apply_mappers(tsg_og_gene_names, ncbi_symbols, ncbi_synonyms, verbose = True,handle_duplicates = \"keep\")\n",
    "OG_entrez = set(map(lambda x : query2target[x],OG))\n",
    "TSG_entrez = set(map(lambda x : query2target[x],TSG))\n",
    "hotspots[\"Entrez_gene_ID\"] = hotspots[\"Hugo_Symbol\"].apply(lambda x : query2target[x])\n",
    "hotspots.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Section 2\n",
    "## Feature Mapping\n",
    "\n",
    "All the code above preps out look up tables and reference information. Now, we need to run all the variants from the WES variants from the MOLI (GDSC dataset)\n",
    "\n",
    "We will need to run the parts below against each patient WES file\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load entire WES feature set used by MOLI paper (GDSC dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>Cancer Type</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>cDNA</th>\n",
       "      <th>AA</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Gene_list</th>\n",
       "      <th>Recurrence Filter</th>\n",
       "      <th>Subs</th>\n",
       "      <th>Truncating</th>\n",
       "      <th>inframe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>KARPAS-45</td>\n",
       "      <td>907272</td>\n",
       "      <td>ALL</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>c.842A&gt;G</td>\n",
       "      <td>p.Y281C</td>\n",
       "      <td>missense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Jurkat</td>\n",
       "      <td>998184</td>\n",
       "      <td>ALL</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>c.589G&gt;A</td>\n",
       "      <td>p.A197T</td>\n",
       "      <td>missense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>KARPAS-45</td>\n",
       "      <td>907272</td>\n",
       "      <td>ALL</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>c.571G&gt;A</td>\n",
       "      <td>p.A191T</td>\n",
       "      <td>missense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>HT-115</td>\n",
       "      <td>907289</td>\n",
       "      <td>COAD/READ</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>c.1456G&gt;A</td>\n",
       "      <td>p.D486N</td>\n",
       "      <td>missense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>KM12</td>\n",
       "      <td>905989</td>\n",
       "      <td>COAD/READ</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>c.770C&gt;A</td>\n",
       "      <td>p.T257N</td>\n",
       "      <td>missense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      SAMPLE  COSMIC_ID Cancer Type  Gene       Transcript       cDNA  \\\n",
       "0  KARPAS-45     907272         ALL  A1BG  ENST00000263100   c.842A>G   \n",
       "1     Jurkat     998184         ALL  A1BG  ENST00000263100   c.589G>A   \n",
       "2  KARPAS-45     907272         ALL  A1BG  ENST00000263100   c.571G>A   \n",
       "3     HT-115     907289   COAD/READ  A1BG  ENST00000263100  c.1456G>A   \n",
       "4       KM12     905989   COAD/READ  A1BG  ENST00000263100   c.770C>A   \n",
       "\n",
       "        AA Classification Gene_list Recurrence Filter  Subs  Truncating  \\\n",
       "0  p.Y281C       missense       NaN               NaN   NaN         NaN   \n",
       "1  p.A197T       missense       NaN               NaN   NaN         NaN   \n",
       "2  p.A191T       missense       NaN               NaN   2.0         NaN   \n",
       "3  p.D486N       missense       NaN               NaN   NaN         NaN   \n",
       "4  p.T257N       missense       NaN               NaN   NaN         NaN   \n",
       "\n",
       "   inframe  \n",
       "0      NaN  \n",
       "1      NaN  \n",
       "2      NaN  \n",
       "3      NaN  \n",
       "4      NaN  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wes_df = pd.read_csv('../data/WES_variants.csv')\n",
    "wes_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>AA</th>\n",
       "      <th>Classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>907272</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.Y281C</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>998184</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.A197T</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>907272</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.A191T</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   COSMIC_ID  Gene       AA Classification\n",
       "0     907272  A1BG  p.Y281C       missense\n",
       "1     998184  A1BG  p.A197T       missense\n",
       "2     907272  A1BG  p.A191T       missense"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapped_wes_df  = wes_df[[\"COSMIC_ID\",\"Gene\",\"AA\",\"Classification\"]]\n",
    "mapped_wes_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kbloom/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:376: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = _infer_fill_value(value)\n",
      "/Users/kbloom/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:494: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[item] = s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mean 485.7 median 288.0\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABIQAAAE/CAYAAAA35xgnAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nOzde5xdVXn4/88DCQnIJUACDQRMsOFmQrhMIJWCQRSVIOFrQbmICUIiBVqsRYnVL6b89NfYogiK0ACWgNwkVkgBrVxFK7cBA0QuEnBKYhBigAiEAIHn+8fZEw+TmcnczpyZ2Z/36zWvs/faa+39rHPOJDtP1lo7MhNJkiRJkiSVxwb1DkCSJEmSJEm9y4SQJEmSJElSyZgQkiRJkiRJKhkTQpIkSZIkSSVjQkiSJEmSJKlkTAhJkiRJkiSVjAkhlU5ENEXEB+t07W0j4q6IeDkivlmPGDojIg6IiCfqHcdAExGXRcTXutj2lYjYqadjkiRJklQuJoSk3jUT+COweWb+Yy0vFBGTI2JpJ9tkRPxl835m/iIzd+n56NRVmblpZj7dlbYRsVFEnBURT0TEqxHx+4j4SUQcUlWnKSJeK5KWL0XEryLi5IjYoKrOqIj4UUT8MSJWRsQjETG9xXVmR8STxXWaIuL7ETG6G12XJEmS1INMCEldFBGDutDs3cCjmZk9HY/eKSr8M+6d5gNTgU8DWwJjgPOAKS3qfSwzN6PyfZ0DnAlcWnX8CmBJcXzr4nzPtbjO4cCxwBbABOAB4OCe7Y4kSZKkrvIfS+oTihEEZ0TEw8WIg2sjYmhxbHpE/LJF/bUjWYrpN98rRjq8EhH/ExF/ERHfjogXI+LxiNirxSUnRsSjxfH/aL5Wcb7DImJh1eiIPVrEeWZEPAy82lpSKCLeFxH3F/24PyLe1xwnMA34YhHnOtPWOtuXliN6mqciRcS7gJ8A2xXneSUitouIfSPi7qJvz0bEdyNio6LtXcVpHirqf7LlKKOI2C0i7iza/yYiDm9x7Qsi4qZidMm9EfGe4lhExLkR8XzxvjwcEePa+C7cGRH/EhH3FXVviIitqo5PKj6XlyLioYiY3KLt1yPif4BVwDpTq4rP7/dFjE9ExMFFeZvvTdV7fUox6uXliPj/IuI9RZs/RcQPq97LyRGxNCL+KSqjaJoi4rjW+lvUb/M710rdlt/9Vt/zVtp9EPgQMDUz783MN4qfn2bm6a21ycyVmbkA+CQwreozmwhclpmvZuaazPx1Zv6klevcXxxfmZkXZOalrV1HkiRJUu8zIaS+5BPAR6iMWtgDmN7Jtl8BhgOvA3cDDxb784Fvtah/HPBh4D3AzkVbImJv4PvAZ6mMfPh3YEFEDKlqewyVERXDMnNN9UmLxMVNwPlF+28BN0XE1pk5HbgS+Ndi2s+tPdSXdWTmq8BHgWXFtTbNzGXAW8A/FOf6KyojNk4p2hxYNJ9Q1L+2Rd8GA/8F/AzYBvg74MqIqJ5Sdgzwz1RGnywGvl6UHwIcSOW9HkYlwbCinS58GvgMsB2whsr7SURsT+X9/RqwFXAG8KOIGFHV9ngqU/M2A/63RR92AU4DJhYjYD4MNBWH23xvqnwE2AeYBHwRmEvlu7QDMK7of7O/KM61PZVE4NwW71VzTB35zrWnrfe8pQ8C92Zmp6YRAmTmfcBS4ICi6B7ggog4OiJ2bOU692Xmks5eR5IkSVLvMSGkvuT8zFyWmS9QSTzs2Ym2P87MBzJzNfBjYHVmXp6ZbwHXAi1HCH03M5cU1/o6f/6H/Azg34sRFG9l5jwqSZlJLeJckpmvtRLHFODJzLyiGBlxNfA48LEa9qXDivPeU8TWRCX58P4ONp8EbArMKUaW3A7cyDuTIP+ZmfcVibIr+fNn+CaVBM2uQGTmY5n5bDvXuiIzFxWJrf8LfCIiNgQ+BdycmTdn5tuZeQvQCBxa1fayzPxN0cc3W5z3LWAIsHtEDM7Mpsx8qhPvzTcy80+Z+RtgEfCzzHw6M1dSGZHV8rP5v5n5emb+nEoi6xOt9LUj37n2tPWetzQc+EPzTkRsVYxIWhkRqztwnWVUknAARwG/oPLZ/K4Y3TSxOLY10N5nK0mSJKkPMCGkvuQPVdurqCQfOqp6/ZLXWtlvea7q0Qv/S2UkClTWRPnH4h/KL0XES1RGf2zXRtuWtqPFqJRif/v2w3+HzvalwyJi54i4MSL+EBF/Av5/KomCjtgOWJKZb1eVtexbq59hkTz6LnAB8FxEzI2Izdu5VsvPZ3AR57uBo1p8Pn8NjGyj7Ttk5mLgc8Bs4PmIuCYitoMOvzed+WxeLBJa1f2o/h4168h3rj0d/b1ZQdX7lJkvZOYwKiOeOjIaaXvghaLti5k5KzPfC2wLLASuj4hoeR1JkiRJfZMJIfUHrwKbNO9ExF/0wDl3qNrekcroB6gkE76emcOqfjYpRvo0a29B6GVU/oFfbUfg992OuHWrqHpvqExTatZanBdSGbE0NjM3B/4JiA5eaxmwQ7xzoeYO9y0zz8/MfYD3Upk69oV2qrf8fN6k8nS2JVRGD1V/Pu/KzDnVl1pPHFdl5l9T+ZwS+EZxqDvvTWu2jMpaTtX9WNZKvY5853rCbVTWzhrV2YbF6J/tgV+2PJaZfwTOoZLA2gq4Fdi3K9eRJEmS1HtMCKk/eAh4b0TsGZXFn2f3wDlPjcqjs7ei8g//5vVyLgZOjoj9ioWQ3xURUyJisw6e92Zg54g4NiIGRcQngd2pTK2qhYXAsRGxYUR8hHdOcXoO2Doitqgq2wz4E/BKROwK/G2L8z1HKwsxF+6lkpz7YkQMLhZz/hhwzfqCjIiJxXs6uDjHairTt9ryqYjYPSI2Ac4G5hdT5n4AfCwiPlz0eWixgHOHkg8RsUtEfKBYn2c1lVE9zXGs773pin+OyiPYDwAOA65rpU53v3Mdkpk/A+6gMpJnvyKuwbQzNS0iNo+Iw6h8xj/IzEeK8m9ExLjiO74ZlfdqcWauKNbGugX4cUTs01wnKo+u/0xP9kmSJElS15kQUp+Xmb+lkhS4FXiSVkYpdMFVVBZHfrr4+VpxrUYqa7p8F3iRyiK90zsR6woq//D/RypTZ74IHFaMoqiF06kkZV6isrjx9VWxPA5cDTxdTEXajsoizMcCL1NJRFzb4nyzgXlF/Xesd5OZb1B5lPhHqYzW+R7w6eI667N5cb0XqUydWkFlVElbrgAuozIdaijw90UMS6g8Nv2fgOVURtd8gY7/WTaEymPU/1ice5viXLD+96az/kClv8uorO1zcmvvVXe/c530cSrJyR9Q+c78jsr35iMt6v1XRLxM5f39MpWFzE+oOr4JlfWtXqLy+/NuKt+NZkdSSY5eC6ykst5SA5XfYUmSJEl9QGS2O7tCknpVRNxJZTTKJfWOpauK0VM/yEynTUmSJEnqkxwhJEmSJEmSVDImhCRJkiRJkkrGKWOSJEmSJEkl4wghSZIkSZKkkjEhJEmSJEmSVDKD6h0AwPDhw3P06NH1DkOSJNXIAw888MfMHFHvOPRO3oNJkjSwtXcP1icSQqNHj6axsbHeYUiSpBqJiP+tdwxal/dgkiQNbO3dgzllTJIkSZIkqWRMCEmSJPVBEbFLRCys+vlTRHwuIraKiFsi4snidcuifkTE+RGxOCIejoi9690HSZLUd5kQkiRJ6oMy84nM3DMz9wT2AVYBPwZmAbdl5ljgtmIf4KPA2OJnJnBh70ctSZL6iz6xhpAkSZLadTDwVGb+b0RMBSYX5fOAO4EzganA5ZmZwD0RMSwiRmbms/UIWJLUNW+++SZLly5l9erV9Q5F/cjQoUMZNWoUgwcP7nAbE0KSJEl939HA1cX2ts1Jnsx8NiK2Kcq3B5ZUtVlalJkQkqR+ZOnSpWy22WaMHj2aiKh3OOoHMpMVK1awdOlSxowZ0+F2ThmTJEnqwyJiI+Bw4Lr1VW2lLFs538yIaIyIxuXLl/dEiJKkHrR69Wq23nprk0HqsIhg66237vSoMhNCkiRJfdtHgQcz87li/7mIGAlQvD5flC8FdqhqNwpY1vJkmTk3Mxsys2HEiBE1DFuS1FUmg9RZXfnOmBCSJEnq247hz9PFABYA04rtacANVeWfLp42NglY6fpBkiSpLa4hJEmS1EdFxCbAh4DPVhXPAX4YEScCzwBHFeU3A4cCi6k8keyEXgxVklQjo2fd1KPna5ozpd3jL730EldddRWnnHJKj163oy677DIOOeQQtttuOwBOOukkPv/5z7P77rvXJZ6Wmpqa+NWvfsWxxx7bqXqNjY1cfvnlnH/++b0RZoc4QkiSJKmPysxVmbl1Zq6sKluRmQdn5tji9YWiPDPz1Mx8T2aOz8zG+kUuSeqvXnrpJb73ve+tU/7WW2/1yvUvu+wyli3784znSy65pM8kg6CS6Lnqqqs6Xa+hoaFPJYPAhJAkSZIkSSrMmjWLp556ij333JOJEydy0EEHceyxxzJ+/HiampoYN27c2rrnnHMOs2fPBmDy5MmceeaZ7Lvvvuy888784he/ACqJpDPOOIPx48ezxx578J3vfAeAs88+m4kTJzJu3DhmzpxJZjJ//nwaGxs57rjj2HPPPXnttdeYPHkyjY2V/+PYdNNN+fKXv8yECROYNGkSzz1XWV7vqaeeYtKkSUycOJGzzjqLTTfdtM3+3Xnnnbz//e/nE5/4BDvvvDOzZs3iyiuvZN9992X8+PE89dRTAEyfPp358+evbdd8zlmzZvGLX/yCPffck3PPPZempiYOOOAA9t57b/bee29+9atftVrvzjvv5LDDDgPghRde4IgjjmCPPfZg0qRJPPzwwwDMnj2bz3zmM0yePJmddtppbQLp1VdfZcqUKUyYMIFx48Zx7bXXduMT/jOnjNVBR4b8rW8YnyRJUp82e4sO1Fm5/jqSpF41Z84cFi1axMKFC7nzzjuZMmUKixYtYsyYMTQ1NbXbds2aNdx3333cfPPN/PM//zO33norc+fO5Xe/+x2//vWvGTRoEC+88AIAp512GmeddRYAxx9/PDfeeCNHHnkk3/3udznnnHNoaGhY5/yvvvoqkyZN4utf/zpf/OIXufjii/nKV77C6aefzumnn84xxxzDRRddtN4+PvTQQzz22GNstdVW7LTTTpx00kncd999nHfeeXznO9/h29/+drvvzznnnMONN94IwKpVq7jlllsYOnQoTz75JMcccwyNjY3r1LvzzjvXnuOrX/0qe+21F9dffz233347n/70p1m4cCEAjz/+OHfccQcvv/wyu+yyC3/7t3/LT3/6U7bbbjtuuqmSS1i5smf+/jQhJEmSpB43enUHhtPXPgxJUjftu+++jBkzpkN1P/7xjwOwzz77rE0e3XrrrZx88skMGlRJP2y11VYA3HHHHfzrv/4rq1at4oUXXuC9730vH/vYx9o9/0YbbbR2lM0+++zDLbfcAsDdd9/N9ddfD8Cxxx7LGWec0e55Jk6cyMiRIwF4z3vewyGHHALA+PHjueOOOzrU12Zvvvkmp512GgsXLmTDDTfkt7/97Xrb/PKXv+RHP/oRAB/4wAdYsWLF2iTPlClTGDJkCEOGDGGbbbbhueeeY/z48ZxxxhmceeaZHHbYYRxwwAGdirEtThmTJEmSJEmtete73rV2e9CgQbz99ttr91evXv2OukOGDAFgww03ZM2aNQBk5jqPRF+9ejWnnHIK8+fP55FHHmHGjBnrnKs1gwcPXnuu6mt0VnOcABtssMHa/Q022GDtOav7mpm88cYbrZ7r3HPPZdttt+Whhx6isbGxzXrVMnOdsuZ+VcfW3Medd96ZBx54gPHjx/OlL32Js88+u4M9bZ8JIUmSJEmSBMBmm23Gyy+/3Oqxbbfdlueff54VK1bw+uuvr50O1Z5DDjmEiy66aG2i5YUXXlib/Bk+fDivvPLKO9bqae/6bZk0adLaETfXXHNNp9q2ZfTo0TzwwAMA3HDDDbz55putxrdy5UpGjhzJBhtswBVXXLF28e32+nHggQdy5ZVXApWpZMOHD2fzzTdvM5Zly5axySab8KlPfYozzjiDBx98sEf66JQxSZIkSZL6qN5eX3brrbdm//33Z9y4cWy88cZsu+22a48NHjyYs846i/32248xY8aw6667rvd8J510Er/97W/ZY489GDx4MDNmzOC0005jxowZjB8/ntGjRzNx4sS19adPn87JJ5/MxhtvzN13392hmL/97W/zqU99im9+85tMmTKFLbbowDp26zFjxgymTp3Kvvvuy8EHH7x2pNQee+zBoEGDmDBhAtOnT+eUU07hb/7mb7juuus46KCD2qy31157rT337NmzOeGEE9hjjz3YZJNNmDdvXruxPPLII3zhC19ggw02YPDgwVx44YXd7h9AtDZUqbc1NDRk86rhZeCi0pKksomIBzJz3dUhVVe1vAfzfkeSuuaxxx5jt912q3cY/cqqVavYeOONiQiuueYarr76am644YZ6h9XrWvvutHcP5gghSZIkSZLUbz3wwAOcdtppZCbDhg3j+9//fr1D6hdMCEmSJEmSpH7rgAMO4KGHHnpH2SOPPMLxxx//jrIhQ4Zw77339mZofZoJIUmSJEmSNKCMHz+ehQsX1juMPs2njEmSJEmSJJWMCSFJkiRJkqSSMSEkSZIkSZJUMiaEJEmSJEmSSsZFpSVJkiRJ6qtmb9HD51vZs+cDTjrpJD7/+c+z++67t1nn+uuvZ+edd263TndddtllHHLIIWy33XadqteR+AciRwhJkiRJkqQuu+SSS9abTLn++ut59NFHaxrHZZddxrJlyzpdryPxD0QmhCRJkiRJ0lpNTU3suuuuTJs2jT322IMjjzySVatWcdttt7HXXnsxfvx4PvOZz/D6668DMHnyZBobGwHYdNNN+fKXv8yECROYNGkSzz33HL/61a9YsGABX/jCF9hzzz156qmnWr3u5MmT+Yd/+AcOPPBAdtttN+6//34+/vGPM3bsWL7yla+sjW3cuHFr25xzzjnMnj2b+fPn09jYyHHHHceee+7Ja6+9xtlnn83EiRMZN24cM2fOJDNbrVcd/9VXX8348eMZN24cZ5555trrtNYvgOuuu45x48YxYcIEDjzwwJ7/MGrIhJAkSZIkSXqHJ554gpkzZ/Lwww+z+eab861vfYvp06dz7bXX8sgjj7BmzRouvPDCddq9+uqrTJo0iYceeogDDzyQiy++mPe9730cfvjh/Nu//RsLFy7kPe95T5vX3Wijjbjrrrs4+eSTmTp1KhdccAGLFi3isssuY8WKFW22O/LII2loaODKK69k4cKFbLzxxpx22mncf//9LFq0iNdee40bb7yx1XrNli1bxplnnsntt9/OwoULuf/++7n++uvb7BfA2WefzX//93/z0EMPsWDBgq6+3XVhQkiSJEmSJL3DDjvswP777w/Apz71KW677TbGjBnDzjvvDMC0adO466671mm30UYbcdhhhwGwzz770NTU1KnrHn744QCMHz+e9773vYwcOZIhQ4aw0047sWTJkk6d64477mC//fZj/Pjx3H777fzmN79pt/7999/P5MmTGTFiBIMGDeK4445b28e2+rX//vszffp0Lr74Yt56661OxVdv600IRcT3I+L5iFhUVfZvEfF4RDwcET+OiGFVx74UEYsj4omI+HCtApckSZIkSbUREV1qN3jw4LVtN9xwQ9asWdOp9kOGDAFggw02WLvdvL9mzRoGDRrE22+/vbZ89erVrZ5n9erVnHLKKcyfP59HHnmEGTNmtFm3WWa2eaytfl100UV87WtfY8mSJey5557tjmLqazoyQugy4CMtym4BxmXmHsBvgS8BRMTuwNHAe4s234uIDXssWkmSJEmSVHPPPPMMd999N1BZV+eDH/wgTU1NLF68GIArrriC97///R0+32abbcbLL7/c7bi23XZbnn/+eVasWMHrr7/OjTfe2Oo1mpM/w4cP55VXXmH+/PnrjWW//fbj5z//OX/84x956623uPrqq9fbx6eeeor99tuPs88+m+HDh3d6FFM9rfex85l5V0SMblH2s6rde4Aji+2pwDWZ+Trwu4hYDOwL3N0j0UqSJEmSVCY1eEx8R+y2227MmzePz372s4wdO5bzzjuPSZMmcdRRR7FmzRomTpzIySef3OHzHX300cyYMYPzzz+f+fPnt7uOUHsGDx7MWWedxX777ceYMWPYdddd1x6bPn06J598MhtvvDF33303M2bMYPz48YwePZqJEye2Wa/ZyJEj+Zd/+RcOOuggMpNDDz2UqVOnthvPF77wBZ588kkyk4MPPpgJEyZ0qV/1EO0NiVpbqZIQujEzx7Vy7L+AazPzBxHxXeCezPxBcexS4CeZOb9lu2oNDQ3ZvKJ3GYyeddN66zTNmdILkUiS1Dsi4oHMbKh3HHqnWt6Deb8jSV3z2GOPsdtuu9U1hqamJg477DAWLVq0/srqM1r77rR3D9atRaUj4svAGuDK5qJWqrWacYqImRHRGBGNy5cv704YkiRJkiRJ6oT1ThlrS0RMAw4DDs4/DzNaCuxQVW0UsKy19pk5F5gLlf+d6mockiRJkiSp54wePbqmo4NOPfVU/ud//ucdZaeffjonnHBCza6pdXUpIRQRHwHOBN6fmauqDi0AroqIbwHbAWOB+7odpSRJkiRJGhAuuOCCeocgOpAQioirgcnA8IhYCnyVylPFhgC3FI9duyczT87M30TED4FHqUwlOzUz36pV8JIkSZIkDTSZ2eXHvqucOrI+dEsdecrYMa0UX9pO/a8DX+90JJIkSZIkldzQoUNZsWIFW2+9tUkhdUhmsmLFCoYOHdqpdl1eQ0iSJEmSJPWsUaNGsXTpUnz4kjpj6NChjBo1qlNtTAhJkiT1URExDLgEGEflya2fAZ4ArgVGA03AJzLzxaj8N/J5wKHAKmB6Zj5Yh7AlSd0wePBgxowZU+8wVALdeuy8JEmSauo84KeZuSswAXgMmAXclpljgduKfYCPUnmgx1hgJnBh74crSZL6CxNCkiRJfVBEbA4cSLF2Y2a+kZkvAVOBeUW1ecARxfZU4PKsuAcYFhEjezlsSZLUT5gQkiRJ6pt2ApYD/xERv46ISyLiXcC2mfksQPG6TVF/e2BJVfulRZkkSdI6TAhJkiT1TYOAvYELM3Mv4FX+PD2sNa09imadZ9BGxMyIaIyIRhcslSSpvEwISZIk9U1LgaWZeW+xP59Kgui55qlgxevzVfV3qGo/CljW8qSZOTczGzKzYcSIETULXpIk9W0mhCRJkvqgzPwDsCQidimKDgYeBRYA04qyacANxfYC4NNRMQlY2Ty1TJIkqSUfOy9JktR3/R1wZURsBDwNnEDlP/R+GBEnAs8ARxV1b6byyPnFVB47f0LvhytJkvoLE0J91OhZN3WoXtOcKTWORJIk1UtmLgQaWjl0cCt1Ezi15kFJkqQBwSljkiRJkiRJJWNCSJIkSZIkqWRMCEmSJEmSJJWMCSFJkiRJkqSSMSEkSZIkSZJUMiaEJEmSJEmSSsaEkCRJkiRJUsmYEJIkSZIkSSoZE0KSJEmSJEklY0JIkiRJkiSpZEwISZIkSZIklYwJIUmSJEmSpJIxISRJkiRJklQyJoQkSZIkSZJKZlC9A1DtjZ5103rrNM2Z0guRSJIkSZKkvsARQpIkSZIkSSVjQkiSJEmSJKlkTAhJkiRJkiSVjGsI9XMdWR9IkiRJkiSpmiOEJEmSJEmSSma9CaGI+H5EPB8Ri6rKtoqIWyLiyeJ1y6I8IuL8iFgcEQ9HxN61DF6SJEmSJEmd15ERQpcBH2lRNgu4LTPHArcV+wAfBcYWPzOBC3smTEmSJEmSJPWU9SaEMvMu4IUWxVOBecX2POCIqvLLs+IeYFhEjOypYCVJkiRJktR9XV1DaNvMfBageN2mKN8eWFJVb2lRJkmSJEmSpD6ipxeVjlbKstWKETMjojEiGpcvX97DYUiSJEmSJKktXU0IPdc8Fax4fb4oXwrsUFVvFLCstRNk5tzMbMjMhhEjRnQxDEmSJEmSJHVWVxNCC4BpxfY04Iaq8k8XTxubBKxsnlomSZIkSZKkvmHQ+ipExNXAZGB4RCwFvgrMAX4YEScCzwBHFdVvBg4FFgOrgBNqELMkSZIkSZK6Yb0Jocw8po1DB7dSN4FTuxuUJEmSJEmSaqenF5WWJElSD4mIpoh4JCIWRkRjUbZVRNwSEU8Wr1sW5RER50fE4oh4OCL2rm/0kiSpL1vvCCF1zuhZN9U7BEmSNLAclJl/rNqfBdyWmXMiYlaxfybwUWBs8bMfcGHxKkmStA5HCEmSJPUvU4F5xfY84Iiq8suz4h5gWPNTYSVJkloyISRJktR3JfCziHggImYWZds2P8W1eN2mKN8eWFLVdmlRJkmStA6njEmSJPVd+2fmsojYBrglIh5vp260UpbrVKoklmYC7Ljjjj0TpSRJ6nccISRJktRHZeay4vV54MfAvsBzzVPBitfni+pLgR2qmo8ClrVyzrmZ2ZCZDSNGjKhl+JIkqQ8zISRJktQHRcS7ImKz5m3gEGARsACYVlSbBtxQbC8APl08bWwSsLJ5apkkSVJLThmTJEnqm7YFfhwRULlnuyozfxoR9wM/jIgTgWeAo4r6NwOHAouBVcAJvR+yJEnqL0wISZIk9UGZ+TQwoZXyFcDBrZQncGovhCZJkgYAE0ICYPSsm9Zbp2nOlF6IRJIkSZIk1ZprCEmSJEmSJJWMCSFJkiRJkqSSMSEkSZIkSZJUMiaEJEmSJEmSSsaEkCRJkiRJUsmYEJIkSZIkSSoZE0KSJEmSJEklY0JIkiRJkiSpZEwISZIkSZIklYwJIUmSJEmSpJIxISRJkiRJklQyJoQkSZIkSZJKxoSQJEmSJElSyZgQkiRJkiRJKhkTQpIkSZIkSSVjQkiSJEmSJKlkTAhJkiRJkiSVjAkhSZIkSZKkkjEhJEmSJEmSVDImhCRJkiRJkkrGhJAkSZIkSVLJdCshFBH/EBG/iYhFEXF1RAyNiDERcW9EPBkR10bERj0VrCRJkiRJkrqvywmhiNge+HugIRnN3wMAABrISURBVDPHARsCRwPfAM7NzLHAi8CJPRGoJElSfxUR4+odgyRJUrXuThkbBGwcEYOATYBngQ8A84vj84AjunkNSZKk/u6iiLgvIk6JiGH1DkaSJKnLCaHM/D1wDvAMlUTQSuAB4KXMXFNUWwps390gJUmS+rPM/GvgOGAHoDEiroqID9U5LEmSVGLdmTK2JTAVGANsB7wL+GgrVbON9jMjojEiGpcvX97VMCRJkvqFzHwS+ApwJvB+4PyIeDwiPl7fyCRJUhl1Z8rYB4HfZebyzHwT+E/gfcCwYgoZwChgWWuNM3NuZjZkZsOIESO6EYYkSVLfFhF7RMS5wGNUptd/LDN3K7bPrWtwkiSplLqTEHoGmBQRm0REAAcDjwJ3AEcWdaYBN3QvREmSpH7vu8CDwITMPDUzHwTIzGVURg1JkiT1qu6sIXQvlcWjHwQeKc41l8ow6M9HxGJga+DSHohTkiSpPzsUuCozXwOIiA0iYhOAzLyivYYRsWFE/Doibiz2x0TEvRHxZERcGxEbFeVDiv3FxfHRNe2RJEnq17r1lLHM/Gpm7pqZ4zLz+Mx8PTOfzsx9M/MvM/OozHy9p4KVJEnqp24FNq7a36Qo64jTqUw1a/YN4NzMHAu8CJxYlJ8IvJiZf0llGto3uhWxJEka0Lr72HlJkiSt39DMfKV5p9jeZH2NImIUMAW4pNgPKusOzS+qzAOOKLanFvsUxw8u6kuSJK3DhJAkSVLtvRoRezfvRMQ+wGsdaPdt4IvA28X+1sBLmbmm2F8KbF9sbw8sASiOryzqS5IkrWPQ+qtIkiSpmz4HXBcRzU9fHQl8sr0GEXEY8HxmPhARk5uLW6maHThWfd6ZwEyAHXfccf2RS5KkAcmEkCRJUo1l5v0RsSuwC5XEzeOZ+eZ6mu0PHB4RhwJDgc2pjBgaFhGDilFAo4DmJNNSYAdgaUQMArYAXmgllrlUHgRCQ0PDOgkjSZJUDk4ZkyRJ6h0TgT2AvYBjIuLT7VXOzC9l5qjMHA0cDdyemccBdwBHFtWmATcU2wuKfYrjt2emCR9JktQqRwhJkiTVWERcAbwHWAi8VRQncHkXTncmcE1EfA34NXBpUX4pcEVELKYyMujobgUtSZIGNBNCkiRJtdcA7N7VETuZeSdwZ7H9NLBvK3VWA0d1PURJklQmThmTJEmqvUXAX9Q7CEmSpGaOEJIkSaq94cCjEXEf8HpzYWYeXr+QJElSmZkQkiRJqr3Z9Q5AkiSpmgkhSZKkGsvMn0fEu4GxmXlrRGwCbFjvuCRJUnm5hpAkSVKNRcQMYD7w70XR9sD19YtIkiSVnQkhSZKk2jsV2B/4E0BmPglsU9eIJElSqZkQkiRJqr3XM/ON5p2IGAR06RH0kiRJPcGEkCRJUu39PCL+Cdg4Ij4EXAf8V51jkiRJJWZCSJIkqfZmAcuBR4DPAjcDX6lrRJIkqdR8ypgkSVKNZebbwMXFjyRJUt2ZEJIkSaqxiPgdrawZlJk71SEcSZIkE0KSJEm9oKFqeyhwFLBVnWKRJElyDSFJkqRay8wVVT+/z8xvAx+od1ySJKm8HCEkSZJUYxGxd9XuBlRGDG1Wp3AkSZJMCEmSJPWCb1ZtrwGagE/UJxRJkiQTQpIkSTWXmQfVOwZJkqRqJoQkSZJqLCI+397xzPxWb8UiSZIEJoQkSZJ6QwMwEVhQ7H8MuAtYUreIJElSqZkQkiRJqr3hwN6Z+TJARMwGrsvMk+oalSRJKi0fOy9JklR7OwJvVO2/AYyuTyiSJEmOEJIkSeoNVwD3RcSPgQT+D3B5fUOSJEllZkJIkiSpxjLz6xHxE+CAouiEzPx1PWOSJEnl5pQxSZKk3rEJ8KfMPA9YGhFj6h2QJEkqr24lhCJiWETMj4jHI+KxiPiriNgqIm6JiCeL1y17KlhJkqT+KCK+CpwJfKkoGgz8oH4RSZKksuvuCKHzgJ9m5q7ABOAxYBZwW2aOBW4r9iVJksrs/wCHA68CZOYyYLO6RiRJkkqtywmhiNgcOBC4FCAz38jMl4CpwLyi2jzgiO4GKUmS1M+9kZlJZUFpIuJddY5HkiSVXHdGCO0ELAf+IyJ+HRGXFDc322bmswDF6zY9EKckSVJ/9sOI+HdgWETMAG4FLq5zTJIkqcS6kxAaBOwNXJiZe1EZAt3h6WERMTMiGiOicfny5d0IQ5IkqW/LzHOA+cCPgF2AszLzO/WNSpIklVl3Hju/FFiamfcW+/OpJISei4iRmflsRIwEnm+tcWbOBeYCNDQ0ZDfikCRJ6rMiYkPgvzPzg8At9Y5HkiQJupEQysw/RMSSiNglM58ADgYeLX6mAXOK1xt6JFLV3ehZN623TtOcKb0QiSRJ/UdmvhURqyJii8xc2dF2ETEUuAsYQuWebX5mfrV4XP01wFbAg8DxmflGRAwBLgf2AVYAn8zMph7ujiRJGiC6M0II4O+AKyNiI+Bp4AQq09B+GBEnAs8AR3XzGpIkSf3dauCRiLiF4kljAJn59+20eR34QGa+EhGDgV9GxE+AzwPnZuY1EXERcCJwYfH6Ymb+ZUQcDXwD+GSN+iNJkvq5biWEMnMh0NDKoYO7c15JkqQB5qbip8OKp5K9UuwOLn4S+ABwbFE+D5hNJSE0tdiGylT+70ZEFOeRJEl6h+6OEJIkSVIbImLHzHwmM+d1sf2GwAPAXwIXAE8BL2XmmqLKUmD7Ynt7YAlAZq6JiJXA1sAfW5xzJjATYMcdd+xKWJIkaQDozlPGJEmS1L7rmzci4kedbZyZb2XmnsAoYF9gt9aqNV+inWPV55ybmQ2Z2TBixIjOhiRJkgYIE0KSJEm1U52k2amrJ8nMl4A7gUnAsIhoHuU9ClhWbC8FdgAojm8BvNDVa0qSpIHNhJAkSVLtZBvb6xURIyJiWLG9MfBB4DHgDuDIolr1E10XFPsUx293/SBJktQW1xCSJEmqnQkR8ScqI4U2LrYp9jMzN2+n7UhgXrGO0AbADzPzxoh4FLgmIr4G/Bq4tKh/KXBFRCymMjLo6Br0R5IkDRAmhCRJkmokMzfsRtuHgb1aKX+aynpCLctXA0d19XqSJKlcnDImSZIkSZJUMiaEJEmSJEmSSsaEkCRJkiRJUsmYEJIkSZIkSSoZE0KSJEmSJEklY0JIkiRJkiSpZEwISZIkSZIklYwJIUmSJEmSpJIxISRJkiRJklQyJoQkSZIkSZJKxoSQJEmSJElSyZgQkiRJkiRJKhkTQpIkSZIkSSVjQkiSJEmSJKlkTAhJkiRJkiSVjAkhSZIkSZKkkjEhJEmSJEmSVDImhCRJkiRJkkrGhJAkSZIkSVLJmBCSJEmSJEkqGRNCkiRJkiRJJWNCSJIkSZIkqWRMCEmSJEmSJJWMCSFJkiRJkqSSMSEkSZIkSZJUMoO6e4KI2BBoBH6fmYdFxBjgGmAr4EHg+Mx8o7vXUf8wetZNHarXNGdKjSORJEmSJElt6YkRQqcDj1XtfwM4NzPHAi8CJ/bANSRJkiRJktRDupUQiohRwBTgkmI/gA8A84sq84AjunMNSZIkSZIk9azujhD6NvBF4O1if2vgpcxcU+wvBbbv5jUkSZJKJyJ2iIg7IuKxiPhNRJxelG8VEbdExJPF65ZFeUTE+RGxOCIejoi969sDSZLUl3U5IRQRhwHPZ+YD1cWtVM022s+MiMaIaFy+fHlXw5AkSRqo1gD/mJm7AZOAUyNid2AWcFsxPf+2Yh/go8DY4mcmcGHvhyxJkvqL7owQ2h84PCKaqCwi/QEqI4aGRUTzYtWjgGWtNc7MuZnZkJkNI0aM6EYYkiRJA09mPpuZDxbbL1NZs3F7YCqVafnwzun5U4HLs+IeKvdkI3s5bEmS1E90+Sljmfkl4EsAETEZOCMzj4uI64AjqSSJpgE39ECckiRJpRURo4G9gHuBbTPzWagkjSJim6La9sCSqmbNU/ef7b1IO2n2Fh2os7L2cUiSVEI98ZSxls4EPh8Ri6msKXRpDa4hSZJUChGxKfAj4HOZ+af2qrZSts7UfaftS5Ik6MYIoWqZeSdwZ7H9NLBvT5xXkiSpzCJiMJVk0JWZ+Z9F8XMRMbIYHTQSeL4oXwrsUNW81an7mTkXmAvQ0NDQ6lqPkiRp4OuRhJAkSZJ6VkQElZHWj2Xmt6oOLaAyLX8O75yevwA4LSKuAfYDVjZPLeurRq++ar11mmofhiRJpWRCSJIkqW/aHzgeeCQiFhZl/0QlEfTDiDgReAY4qjh2M3AosBhYBZzQu+FKkqT+xISQJElSH5SZv6T1dYEADm6lfgKn1jQoSZI0YNRiUWlJkiRJkiT1YSaEJEmSJEmSSsaEkCRJkiRJUsmYEJIkSZIkSSoZE0KSJEmSJEklY0JIkiRJkiSpZEwISZIkSZIklYwJIUmSJEmSpJIxISRJkiRJklQyJoQkSZIkSZJKxoSQJEmSJElSyZgQkiRJkiRJKplB9Q5AkiRJ6rbZW3SgzsraxyFJUj/hCCFJkiRJkqSSMSEkSZIkSZJUMiaEJEmSJEmSSsaEkCRJkiRJUsm4qHQnjJ51U71DkCRJUle58LQkSWs5QkiSJEmSJKlkHCGkuuip0VZNc6b0yHkkSZIkSSoTRwhJkiRJkiSVjAkhSZIkSZKkkjEhJEmSJEmSVDImhCRJkiRJkkrGhJAkSZIkSVLJ+JQxSZIkqdnsLTpQZ2Xvn0uSpB7mCCFJkiRJkqSS6fIIoYjYAbgc+AvgbWBuZp4XEVsB1wKjgSbgE5n5YvdDlSRJkvqAjoz8kSSpj+vOCKE1wD9m5m7AJODUiNgdmAXclpljgduKfUmSJEmSJPURXU4IZeazmflgsf0y8BiwPTAVmFdUmwcc0d0gJUmSyiYivh8Rz0fEoqqyrSLiloh4snjdsiiPiDg/IhZHxMMRsXf9IpckSf1Bj6whFBGjgb2Ae4FtM/NZqCSNgG164hqSJEklcxnwkRZlbY3E/igwtviZCVzYSzFKkqR+qtsJoYjYFPgR8LnM/FMn2s2MiMaIaFy+fHl3w5AkSRpQMvMu4IUWxW2NxJ4KXJ4V9wDDImJk70QqSZL6o24lhCJiMJVk0JWZ+Z9F8XPNNyDF6/Ottc3MuZnZkJkNI0aM6E4YkiRJZdHWSOztgSVV9ZYWZZIkSa3qckIoIgK4FHgsM79VdWgBMK3Yngbc0PXwJEmS1AHRSlm2WtFR2pIkie6NENofOB74QEQsLH4OBeYAH4qIJ4EPFfuSJEnqvrZGYi8FdqiqNwpY1toJHKUtSZIABnW1YWb+ktb/Nwrg4K6eV5IkSW1qHok9h3eOxF4AnBYR1wD7ASubp5ZJkiS1pssJIUmSJNVORFwNTAaGR8RS4KtUEkE/jIgTgWeAo4rqNwOHAouBVcAJvR6wJEnqV0wISZIk9UGZeUwbh9YZiZ2ZCZxa24gkSdJA0u3HzkuSJEmSJKl/cYSQ+rXRs25ab52mOVN6IRJJkiRJkvoPE0KSJEnqu2ZvUe8IJEkakEwISZ3giCRJkiRJ0kDgGkKSJEmSJEklY0JIkiRJkiSpZJwyJkmSJNVLR9ZImr2y9nFIkkrHEUKSJEmSJEklY0JIkiRJkiSpZJwypgGvI08GA58OJkmSJEkqD0cISZIkSZIklYwJIUmSJEmSpJIxISRJkiRJklQyriEkFTq61pAkSZIkSf2dI4QkSZIkSZJKxhFCUh10ZDSSTz2TJEkAzN6iA3VW1j4OSdKA4gghSZIkSZKkknGEEK4dI0mSpH7OUUSSpE5yhJAkSZIkSVLJOEJI6mGOOJMkSZIk9XWOEJIkSZIkSSoZRwhJUgn4ZDtJ/dXo1Vd1qF7T0GNrHIkkSQOLI4QkSZIkSZJKxhFCUh/V0bWIOjKqo7dHhzgaRZKkfswnlklSKZgQkiRJktTzTCxJUp9mQkiSVAqOXJNUeh1J0EiSSsOEkCRJkqTOMbkkSf2eCSFJQN8cPdEXY5I6ur5XR/j9lXpOR55G5pPIJEn6M58yJkmSJEmSVDI1GyEUER8BzgM2BC7JzDm1upZUZj05WqE3r9WbcXf0ev11tEZ/fS978zwd1Rffy47oi08bVH14/9V9jjTqZT059cwFqiWpU2qSEIqIDYELgA8BS4H7I2JBZj5ai+tJkiSVnfdf69eRZI96jsk1+aQ5qW+r1QihfYHFmfk0QERcA0wFvCGR1KP68+iQgcz3snf5fqvg/Vcv6WhiqaeSHSayOqinRht1JEHR24tq92RMZU/A9GSSqq8lvPwOqJNqlRDaHlhStb8U2K9G15IkSZL3X32OiZz168n3qCMJuA6NWurAP6p7+7PtSEwd1aEpxB1JZnYgqdBT73ev69GpjH2vfz32HehtPZQY7dD3soPT6Hvz96kWIjN7/qQRRwEfzsyTiv3jgX0z8++q6swEZha7uwBP9HggFcOBP9bo3H2J/Rw4ytBHKEc/y9BHsJ8DSS37+O7MHFGjc4uO3X8V5d6D9awy9LMMfQT7OZCUoY9Qjn6WoY9Qp3uwWo0QWgrsULU/ClhWXSEz5wJza3T9tSKiMTMban2derOfA0cZ+gjl6GcZ+gj2cyApQx8HuPXef4H3YD2tDP0sQx/Bfg4kZegjlKOfZegj1K+ftXrs/P3A2IgYExEbAUcDC2p0LUmSJHn/JUmSOqEmI4Qyc01EnAb8N5XHnn4/M39Ti2tJkiTJ+y9JktQ5tZoyRmbeDNxcq/N3Qs2HRPcR9nPgKEMfoRz9LEMfwX4OJGXo44DWh+6/oDzfpzL0swx9BPs5kJShj1COfpahj1CnftZkUWlJkiRJkiT1XbVaQ0iSJEmSJEl91IBOCEXERyLiiYhYHBGz6h1PV0XEDhFxR0Q8FhG/iYjTi/LZEfH7iFhY/Bxa1eZLRb+fiIgP1y/6zomIpoh4pOhPY1G2VUTcEhFPFq9bFuUREecX/Xw4Ivaub/QdExG7VH1mCyPiTxHxuf7+eUbE9yPi+YhYVFXW6c8uIqYV9Z+MiGn16Et72ujnv0XE40VffhwRw4ry0RHxWtVnelFVm32K7/ri4r2IevSnLW30s9Pf0b7853Abfby2qn9NEbGwKO/Pn2Vbf4cMuN9P9Q19+fe+s9r5/enXf2e3Jgb4PVgM0Psv8B4sBtA9WBt9HFD3X1COe7B2/v7oW7+bmTkgf6gspvgUsBOwEfAQsHu94+piX0YCexfbmwG/BXYHZgNntFJ/96K/Q4AxxfuwYb370cG+NgHDW5T9KzCr2J4FfKPYPhT4CRDAJODeesffhf5uCPwBeHd//zyBA4G9gUVd/eyArYCni9cti+0t6923DvTzEGBQsf2Nqn6Orq7X4jz3AX9VvAc/AT5a7751oJ+d+o729T+HW+tji+PfBM4aAJ9lW3+HDLjfT3/q/9PXf++70B/vwQbgPRgD6P6riNV7sBwY92Bt9LFT39H+8Odwa/1scbzf34O18/dHn/rdHMgjhPYFFmfm05n5BnANMLXOMXVJZj6bmQ8W2y8DjwHbt9NkKnBNZr6emb8DFlN5P/qrqcC8YnsecERV+eVZcQ8wLCJG1iPAbjgYeCoz/7edOv3i88zMu4AXWhR39rP7MHBLZr6QmS8CtwAfqX30HddaPzPzZ5m5pti9B/h/7d3Li1xFGIbx58MbeImoqEiiOJG4NpJFQONKghENqCAjQkQFEXQhblzM/+BKURAlKFFEVJydggtXIpJojOIl0Y0hwwSyUMGNl3Jx6jg9sXtMT8j0qarnB830FJ1Of1NV57ypqXOyZa33yLVuSil9mroj/eus/GwGYUJ/TjJpjA76OLxWjfk3TA8Cb631HoX05aRzSHXzU4Mw6Hk/LTNYtRmsmvwFZrCaMlgL+QvayGCl5K+aF4Q2Az+PfH+ctU/gRYiIG4HtwGe56em8pey1frsZZdeegI8i4mBEPJHbrk0pLUE3sYBrcnvJdfbmWX2wq60/p+27kmvtPUa3ut+bi4gvIuKTiNiV2zbT1dYrqc5pxmjJ/bkLWE4pHR1pK74vTzuHtDg/de5VO07MYFVlsNrzF7R5jK85g7WSv6DCDDbk/FXzgtC46weL/i/VIuJS4F3gmZTSr8BLwE3ALcAS3dY6KLv221JKtwJ7gKci4o41XltynUTEhcBe4J3cVGN/TjKppqJrjYgF4E/gQG5aAm5IKW0HngXejIhNlFvntGO01DoBHmL1PxaK78sx55CJLx3TVnp/auNUOU7MYP9RbJ2N5y+o9BhfeQZrKX9BZRls6Pmr5gWh48D1I99vAU7M6LOctYi4gG4gHUgpvQeQUlpOKf2VUvobeIWVbazF1p5SOpG/ngTep6tpud+GnL+ezC8vts5sD3AopbQMdfYn0/ddsbXmG7zdAzyct62St/Ceys8P0l3PfTNdnaNbmouocx1jtMj+jIjzgfuBt/u20vty3DmEhuanNlR148QMVl0GayF/QUPH+NozWCv5C+rLYCXkr5oXhD4HtkXEXP5NwDywOOPPtC75OspXgW9TSs+PtI9eq30f0N+lfRGYj4iLImIO2EZ3w61Bi4hLIuKy/jndTeK+pqunv5v6I8AH+fkisC/fkX0n8Eu//a4Qq1a/a+vPbNq++xDYHRFX5O2wu3PboEXEXcBzwN6U0u8j7VdHxHn5+Va6vvsp1/pbROzM83sfKz+bwVrHGC31OHwn8F1K6d9tyCX35aRzCI3MT224Uuf9WGawKjNYC/kLGjnGt5DBGspfUFEGKyZ/pQHcgftcPeju1P0D3Sriwqw/z1nUcTvdtrCvgC/z427gDeBIbl8Erhv5Mwu57u8Z0N3W/6fOrXR3wT8MfNP3GXAV8DFwNH+9MrcH8GKu8wiwY9Y1TFHrxcAp4PKRtqL7ky5cLQF/0K1kP76evqO7/vtYfjw667rOsM5jdNf29vPz5fzaB/JYPgwcAu4deZ8ddCf0H4EXgJh1bWdQ59RjdMjH4XE15vb9wJOnvbbkvpx0DqlufvoYxmPI834dtZjBKspgVJi/8uc0g1WSwSbUWFX+mlRnbt9PJRmMQvJX5L9AkiRJkiRJjaj5kjFJkiRJkiSN4YKQJEmSJElSY1wQkiRJkiRJaowLQpIkSZIkSY1xQUiSJEmSJKkxLghJkiRJkiQ1xgUhSZIkSZKkxrggJEmSJEmS1Jh/AMu0oMenoW7FAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1440x360 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# define functional effect of a mutation \n",
    "gdsc_truncating = ['frameshift','nonsense','stop_lost','ess_splice']\n",
    "gdsc_point = ['missense','Missense','inframe']\n",
    "mapped_wes_df.loc[mapped_wes_df[\"Classification\"].isin(gdsc_truncating),\"function\"] = \"truncating_mutation\"\n",
    "mapped_wes_df.loc[mapped_wes_df[\"Classification\"].isin(gdsc_point),\"function\"] = \"point_mutation\"\n",
    "mapped_wes_df = mapped_wes_df.loc[mapped_wes_df[\"Classification\"].isin(gdsc_truncating+gdsc_point),:]\n",
    "\n",
    "plt.figure(figsize=(20,5))\n",
    "plt.subplot(1,2,1)\n",
    "n_mutations_gdsc = mapped_wes_df.groupby(\"COSMIC_ID\").size().sort_values(ascending = False)\n",
    "tmp = plt.hist(n_mutations_gdsc, bins=50,range = (0,2000))\n",
    "tmp = plt.title(\"number of mutations per sample in GDSC\")\n",
    "ax = plt.subplot(1,2,2)\n",
    "n_LoF = mapped_wes_df.loc[mapped_wes_df[\"Classification\"].isin(gdsc_truncating ),:].groupby(\"COSMIC_ID\").size().sort_values(ascending = False)\n",
    "n_GoF = mapped_wes_df.loc[mapped_wes_df[\"Classification\"].isin(gdsc_point),:].groupby(\"COSMIC_ID\").size().sort_values(ascending = False)\n",
    "df = pd.DataFrame.from_dict({\"truncating_mutations\":n_LoF,\"point_mutations\":n_GoF})\n",
    "#tmp = plt.hist(n_LoF, bins=50)\n",
    "#tmp = plt.title(\"LoF mutations per sample in GDSC\")\n",
    "df.plot.hist(stacked = True, bins = 50,ax =ax,range = (0,2000))\n",
    "print(\"mean\",round(np.mean(n_mutations_gdsc),1), \"median\",np.median(n_mutations_gdsc))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Helper funcs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def binarize(x):\n",
    "    if x==0:\n",
    "        return x\n",
    "    else:\n",
    "        return 1\n",
    "\n",
    "def get_aa_pos_GDSC(x):\n",
    "    l = re.findall(r'\\d+', x)\n",
    "    if len(l) == 0:\n",
    "        #print(x, \"positions not found.\",file = sys.stderr)\n",
    "        return (\"NA\",\"NA\")\n",
    "    elif len(l) == 1:\n",
    "        return (int(l[0]), int(l[0]))\n",
    "    else:\n",
    "        return (int(l[0]), int(l[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get actual position of mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>AA</th>\n",
       "      <th>Classification</th>\n",
       "      <th>function</th>\n",
       "      <th>aa_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>907272</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.Y281C</td>\n",
       "      <td>missense</td>\n",
       "      <td>point_mutation</td>\n",
       "      <td>(281, 281)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>998184</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.A197T</td>\n",
       "      <td>missense</td>\n",
       "      <td>point_mutation</td>\n",
       "      <td>(197, 197)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>907272</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.A191T</td>\n",
       "      <td>missense</td>\n",
       "      <td>point_mutation</td>\n",
       "      <td>(191, 191)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>907289</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.D486N</td>\n",
       "      <td>missense</td>\n",
       "      <td>point_mutation</td>\n",
       "      <td>(486, 486)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>905989</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.T257N</td>\n",
       "      <td>missense</td>\n",
       "      <td>point_mutation</td>\n",
       "      <td>(257, 257)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   COSMIC_ID  Gene       AA Classification        function      aa_pos\n",
       "0     907272  A1BG  p.Y281C       missense  point_mutation  (281, 281)\n",
       "1     998184  A1BG  p.A197T       missense  point_mutation  (197, 197)\n",
       "2     907272  A1BG  p.A191T       missense  point_mutation  (191, 191)\n",
       "3     907289  A1BG  p.D486N       missense  point_mutation  (486, 486)\n",
       "4     905989  A1BG  p.T257N       missense  point_mutation  (257, 257)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapped_wes_df[\"aa_pos\"] = mapped_wes_df[\"AA\"].apply(lambda x : get_aa_pos_GDSC(x))\n",
    "mapped_wes_df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look to see if mutaitons are in onco genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSEMBL gene IDs found: 101\n",
      "Gene symbols: 18989\n"
     ]
    }
   ],
   "source": [
    "ensg_genes = []\n",
    "symbols = []\n",
    "for gene in set(mapped_wes_df[\"Gene\"].values):\n",
    "    if gene.startswith(\"ENSG\"):\n",
    "        ensg_genes.append(gene)\n",
    "        \n",
    "    else:\n",
    "        symbols.append(gene)\n",
    "print(\"ENSEMBL gene IDs found:\",len(ensg_genes))\n",
    "ensg_genes = mapped_wes_df.loc[mapped_wes_df[\"Gene\"].isin(ensg_genes), :]\n",
    "gdsc_hugo =  mapped_wes_df.loc[mapped_wes_df[\"Gene\"].isin(symbols), :]\n",
    "print(\"Gene symbols:\",len(set(gdsc_hugo[\"Gene\"].values)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map to Entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped: 18466 \n",
      "\tdirectly via main_mapper 16773 \n",
      "\tvia alternative mapper 472 \n",
      "\tvia one of multiple synonyms in alternative mapper 1221 \n",
      "\tLOC 0 \n",
      "Unmapped: 523 \n",
      "\trecognized symbols without Entrez ID 0 \n",
      "\tmultiple query_ids map to the same target_id 0 \n",
      "\tquery_ids map to multiple target_ids in the main mapper 0 \n",
      "\tquery_ids map to multiple target_ids in the alternative mapper 64 \n",
      "\tLOC not found in Entrez 36 \n",
      "\tNot found at all: 423\n",
      "Warning: query IDs mapping to duplicated target IDs in mapping table: 168\n",
      "Warning: query IDs not mapped to any target IDs excluded: 523\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IDs mapped to multiple target IDs are kept:\n",
      " [143872, 341019, 27177, 387119, 84530, 26165, 11321, 122945, 150082, 728642, 2122, 79949, 157777, 6241, 100289124, 8805, 127602, 387707, 345222, 84108, 91283, 84631, 144535, 245913, 388761, 4253, 163486, 54438, 283310, 84148, 57018, 284352, 5315, 339669, 23254, 23272, 119016, 340204, 64753, 64760, 25849, 286464, 129285, 4360, 10507, 4883, 5414, 23334, 256815, 23345, 57662, 54617, 84316, 200030, 10592, 57705, 246126, 92017, 126321, 10613, 374, 445815, 79741, 51586, 647060, 284565, 63895, 201625, 29086, 1954, 9127, 55719, 29099, 941, 155066, 140733, 114112, 85452, 139728, 84458, 23028, 54777, 79868]\n"
     ]
    }
   ],
   "source": [
    "# mapping to Entrez\n",
    "df = gdsc_hugo[[\"Gene\"]].copy()\n",
    "df.set_index(\"Gene\",drop= False ,inplace=True)\n",
    "df.drop_duplicates(inplace = True)\n",
    "df, query2target,not_mapped  = apply_mappers(df, ncbi_symbols, ncbi_synonyms, verbose = True,handle_duplicates = \"keep\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map out to get entrez_gene_id \n",
    "\n",
    "### This value, entrez_gene_id is important, it will become the name of all of our feature columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>AA</th>\n",
       "      <th>Classification</th>\n",
       "      <th>function</th>\n",
       "      <th>aa_pos</th>\n",
       "      <th>Entrez_gene_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>907272</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.Y281C</td>\n",
       "      <td>missense</td>\n",
       "      <td>point_mutation</td>\n",
       "      <td>(281, 281)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>998184</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.A197T</td>\n",
       "      <td>missense</td>\n",
       "      <td>point_mutation</td>\n",
       "      <td>(197, 197)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>907272</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.A191T</td>\n",
       "      <td>missense</td>\n",
       "      <td>point_mutation</td>\n",
       "      <td>(191, 191)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>907289</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.D486N</td>\n",
       "      <td>missense</td>\n",
       "      <td>point_mutation</td>\n",
       "      <td>(486, 486)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>905989</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.T257N</td>\n",
       "      <td>missense</td>\n",
       "      <td>point_mutation</td>\n",
       "      <td>(257, 257)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>909698</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.R90H</td>\n",
       "      <td>missense</td>\n",
       "      <td>point_mutation</td>\n",
       "      <td>(90, 90)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>905971</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.T46M</td>\n",
       "      <td>missense</td>\n",
       "      <td>point_mutation</td>\n",
       "      <td>(46, 46)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>909761</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.P41fs*33</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>truncating_mutation</td>\n",
       "      <td>(41, 33)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>1327769</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.E249D</td>\n",
       "      <td>missense</td>\n",
       "      <td>point_mutation</td>\n",
       "      <td>(249, 249)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>1330931</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>p.Y184*</td>\n",
       "      <td>nonsense</td>\n",
       "      <td>truncating_mutation</td>\n",
       "      <td>(184, 184)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   COSMIC_ID  Gene          AA Classification             function  \\\n",
       "0     907272  A1BG     p.Y281C       missense       point_mutation   \n",
       "1     998184  A1BG     p.A197T       missense       point_mutation   \n",
       "2     907272  A1BG     p.A191T       missense       point_mutation   \n",
       "3     907289  A1BG     p.D486N       missense       point_mutation   \n",
       "4     905989  A1BG     p.T257N       missense       point_mutation   \n",
       "5     909698  A1BG      p.R90H       missense       point_mutation   \n",
       "6     905971  A1BG      p.T46M       missense       point_mutation   \n",
       "7     909761  A1BG  p.P41fs*33     frameshift  truncating_mutation   \n",
       "8    1327769  A1BG     p.E249D       missense       point_mutation   \n",
       "9    1330931  A1BG     p.Y184*       nonsense  truncating_mutation   \n",
       "\n",
       "       aa_pos  Entrez_gene_ID  \n",
       "0  (281, 281)               1  \n",
       "1  (197, 197)               1  \n",
       "2  (191, 191)               1  \n",
       "3  (486, 486)               1  \n",
       "4  (257, 257)               1  \n",
       "5    (90, 90)               1  \n",
       "6    (46, 46)               1  \n",
       "7    (41, 33)               1  \n",
       "8  (249, 249)               1  \n",
       "9  (184, 184)               1  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc_hugo = gdsc_hugo.loc[~gdsc_hugo[\"Gene\"].isin(not_mapped),:].copy()\n",
    "gdsc_hugo[\"Entrez_gene_ID\"] = gdsc_hugo[\"Gene\"].apply(lambda x : query2target[x])\n",
    "gdsc_hugo.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Score each variant / mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "w = 0.05\n",
    "\n",
    "\n",
    "\n",
    "gdsc_scores = {}\n",
    "for sample in list(set(gdsc_hugo[\"COSMIC_ID\"].values)):\n",
    "    df = gdsc_hugo.loc[gdsc_hugo[\"COSMIC_ID\"] == sample,:]\n",
    "    gdsc_scores[sample] = mutations2score(df, TSG_entrez,OG_entrez, hotspots, w=w)\n",
    "gdsc_scores = pd.DataFrame.from_dict(gdsc_scores)\n",
    "gdsc_scores.fillna(0,inplace=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save file with contious value scoring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>907268</th>\n",
       "      <th>907269</th>\n",
       "      <th>907270</th>\n",
       "      <th>907271</th>\n",
       "      <th>907272</th>\n",
       "      <th>907273</th>\n",
       "      <th>907274</th>\n",
       "      <th>907275</th>\n",
       "      <th>907276</th>\n",
       "      <th>907277</th>\n",
       "      <th>...</th>\n",
       "      <th>753620</th>\n",
       "      <th>753621</th>\n",
       "      <th>688086</th>\n",
       "      <th>753623</th>\n",
       "      <th>753622</th>\n",
       "      <th>753624</th>\n",
       "      <th>688087</th>\n",
       "      <th>753618</th>\n",
       "      <th>917486</th>\n",
       "      <th>688121</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>7157</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5290</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3845</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4893</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows  1001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      907268  907269  907270  907271  907272  907273  907274  907275  907276  \\\n",
       "7157    -1.0     0.0     0.0    -1.0    -1.0    -1.0    -1.0    -1.0     0.0   \n",
       "5290     0.0     0.0     0.0     0.0     0.2     0.0     0.0     0.0     0.0   \n",
       "3845     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "4893     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "      907277  ...  753620  753621  688086  753623  753622  753624  688087  \\\n",
       "7157    -1.0  ...     0.0    -1.0    -1.0    -1.0    -1.0    -1.0    -1.0   \n",
       "5290     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "3845     0.0  ...     0.0     0.0     0.0     0.0     0.0     1.0     0.0   \n",
       "4893    -1.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "      753618  917486  688121  \n",
       "7157    -1.0    -1.0     0.0  \n",
       "5290     0.0     0.0     0.0  \n",
       "3845     0.0     1.0     0.0  \n",
       "4893     0.0     0.0     0.0  \n",
       "\n",
       "[4 rows x 1001 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc_scores.to_csv(\"../data/GDSC.non_bin_mutations.tsv\",sep = \"\\t\")\n",
    "gdsc_scores.loc[[7157,5290,3845,4893],:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make decsion if we want to binarize the score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>907268</th>\n",
       "      <th>907269</th>\n",
       "      <th>907270</th>\n",
       "      <th>907271</th>\n",
       "      <th>907272</th>\n",
       "      <th>907273</th>\n",
       "      <th>907274</th>\n",
       "      <th>907275</th>\n",
       "      <th>907276</th>\n",
       "      <th>907277</th>\n",
       "      <th>...</th>\n",
       "      <th>753620</th>\n",
       "      <th>753621</th>\n",
       "      <th>688086</th>\n",
       "      <th>753623</th>\n",
       "      <th>753622</th>\n",
       "      <th>753624</th>\n",
       "      <th>688087</th>\n",
       "      <th>753618</th>\n",
       "      <th>917486</th>\n",
       "      <th>688121</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>7157</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5290</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3845</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2312</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4893</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  1001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      907268  907269  907270  907271  907272  907273  907274  907275  907276  \\\n",
       "7157     1.0     0.0     0.0     1.0     1.0     1.0     1.0     1.0     0.0   \n",
       "5290     0.0     0.0     0.0     0.0     1.0     0.0     0.0     0.0     0.0   \n",
       "3845     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "2312     1.0     0.0     0.0     1.0     0.0     1.0     0.0     0.0     0.0   \n",
       "4893     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "      907277  ...  753620  753621  688086  753623  753622  753624  688087  \\\n",
       "7157     1.0  ...     0.0     1.0     1.0     1.0     1.0     1.0     1.0   \n",
       "5290     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "3845     0.0  ...     0.0     0.0     0.0     0.0     0.0     1.0     0.0   \n",
       "2312     0.0  ...     1.0     1.0     0.0     0.0     0.0     1.0     0.0   \n",
       "4893     1.0  ...     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "      753618  917486  688121  \n",
       "7157     1.0     1.0     0.0  \n",
       "5290     0.0     0.0     0.0  \n",
       "3845     0.0     1.0     0.0  \n",
       "2312     0.0     0.0     0.0  \n",
       "4893     0.0     0.0     0.0  \n",
       "\n",
       "[5 rows x 1001 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc_scores = gdsc_scores.applymap(lambda x: binarize(x))\n",
    "gdsc_scores.to_csv(\"../data/GDSC.binary_mutations.tsv\",sep = \"\\t\")\n",
    "gdsc_scores.loc[[7157,5290,3845,2312,4893],:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prep data for transpose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>907268</th>\n",
       "      <th>907269</th>\n",
       "      <th>907270</th>\n",
       "      <th>907271</th>\n",
       "      <th>907272</th>\n",
       "      <th>907273</th>\n",
       "      <th>907274</th>\n",
       "      <th>907275</th>\n",
       "      <th>907276</th>\n",
       "      <th>907277</th>\n",
       "      <th>...</th>\n",
       "      <th>753620</th>\n",
       "      <th>753621</th>\n",
       "      <th>688086</th>\n",
       "      <th>753623</th>\n",
       "      <th>753622</th>\n",
       "      <th>753624</th>\n",
       "      <th>688087</th>\n",
       "      <th>753618</th>\n",
       "      <th>917486</th>\n",
       "      <th>688121</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  1001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         907268  907269  907270  907271  907272  907273  907274  907275  \\\n",
       "Gene_ID                                                                   \n",
       "1           0.0     0.0     0.0     0.0     1.0     0.0     0.0     0.0   \n",
       "2           1.0     0.0     0.0     0.0     1.0     0.0     0.0     0.0   \n",
       "9           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "10          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "12          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "         907276  907277  ...  753620  753621  688086  753623  753622  753624  \\\n",
       "Gene_ID                  ...                                                   \n",
       "1           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "2           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "9           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "10          0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "12          0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "         688087  753618  917486  688121  \n",
       "Gene_ID                                  \n",
       "1           0.0     0.0     0.0     0.0  \n",
       "2           0.0     0.0     0.0     0.0  \n",
       "9           0.0     0.0     0.0     0.0  \n",
       "10          0.0     0.0     0.0     0.0  \n",
       "12          0.0     0.0     0.0     0.0  \n",
       "\n",
       "[5 rows x 1001 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load one back in \n",
    "cosmic_wes_binary = pd.read_csv(\"../data/GDSC.binary_mutations.tsv\",sep = \"\\t\")\n",
    "cosmic_wes_binary.set_index(\"Unnamed: 0\",drop= True ,inplace=True)\n",
    "cosmic_wes_binary.index.names = ['Gene_ID']\n",
    "cosmic_wes_binary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene_ID</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>101060321</th>\n",
       "      <th>101927546</th>\n",
       "      <th>101927722</th>\n",
       "      <th>101928638</th>\n",
       "      <th>102724473</th>\n",
       "      <th>102724928</th>\n",
       "      <th>105375355</th>\n",
       "      <th>105378803</th>\n",
       "      <th>107403068</th>\n",
       "      <th>109731405</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>907268</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>907269</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>907270</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>907271</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>907272</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  18381 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene_ID  1          2          9          10         12         13         \\\n",
       "907268         0.0        1.0        0.0        0.0        0.0        0.0   \n",
       "907269         0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "907270         0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "907271         0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "907272         1.0        1.0        0.0        0.0        0.0        0.0   \n",
       "\n",
       "Gene_ID  14         15         16         18         ...  101060321  \\\n",
       "907268         0.0        0.0        0.0        0.0  ...        0.0   \n",
       "907269         0.0        0.0        0.0        0.0  ...        0.0   \n",
       "907270         0.0        0.0        0.0        0.0  ...        0.0   \n",
       "907271         0.0        0.0        0.0        0.0  ...        0.0   \n",
       "907272         1.0        0.0        1.0        0.0  ...        0.0   \n",
       "\n",
       "Gene_ID  101927546  101927722  101928638  102724473  102724928  105375355  \\\n",
       "907268         0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "907269         0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "907270         0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "907271         0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "907272         0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "\n",
       "Gene_ID  105378803  107403068  109731405  \n",
       "907268         0.0        0.0        0.0  \n",
       "907269         0.0        0.0        0.0  \n",
       "907270         0.0        0.0        0.0  \n",
       "907271         0.0        0.0        0.0  \n",
       "907272         0.0        0.0        0.0  \n",
       "\n",
       "[5 rows x 18381 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cosmic_wes_binary_transposed = cosmic_wes_binary.transpose()\n",
    "cosmic_wes_binary_transposed.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save out file transposed with COSMIC ID as rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "cosmic_wes_binary_transposed.index.names = ['COSMIC_ID']\n",
    "cosmic_wes_binary_transposed.to_csv(\"../data/wes_binary_scored_transposed.tsv\",sep = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene_ID</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>101060321</th>\n",
       "      <th>101927546</th>\n",
       "      <th>101927722</th>\n",
       "      <th>101928638</th>\n",
       "      <th>102724473</th>\n",
       "      <th>102724928</th>\n",
       "      <th>105375355</th>\n",
       "      <th>105378803</th>\n",
       "      <th>107403068</th>\n",
       "      <th>109731405</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>907268</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>907269</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>907270</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>907271</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>907272</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  18381 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene_ID    1          2          9          10         12         13         \\\n",
       "COSMIC_ID                                                                     \n",
       "907268           0.0        1.0        0.0        0.0        0.0        0.0   \n",
       "907269           0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "907270           0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "907271           0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "907272           1.0        1.0        0.0        0.0        0.0        0.0   \n",
       "\n",
       "Gene_ID    14         15         16         18         ...  101060321  \\\n",
       "COSMIC_ID                                              ...              \n",
       "907268           0.0        0.0        0.0        0.0  ...        0.0   \n",
       "907269           0.0        0.0        0.0        0.0  ...        0.0   \n",
       "907270           0.0        0.0        0.0        0.0  ...        0.0   \n",
       "907271           0.0        0.0        0.0        0.0  ...        0.0   \n",
       "907272           1.0        0.0        1.0        0.0  ...        0.0   \n",
       "\n",
       "Gene_ID    101927546  101927722  101928638  102724473  102724928  105375355  \\\n",
       "COSMIC_ID                                                                     \n",
       "907268           0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "907269           0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "907270           0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "907271           0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "907272           0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "\n",
       "Gene_ID    105378803  107403068  109731405  \n",
       "COSMIC_ID                                   \n",
       "907268           0.0        0.0        0.0  \n",
       "907269           0.0        0.0        0.0  \n",
       "907270           0.0        0.0        0.0  \n",
       "907271           0.0        0.0        0.0  \n",
       "907272           0.0        0.0        0.0  \n",
       "\n",
       "[5 rows x 18381 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cosmic_wes_binary_transposed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
